### PR TITLE
new ISM template

### PR DESCRIPTION
**Pull request by bowenlan-amzn**
*Issue #, if available:*

*Description of changes:*
We are gonna remove policyID usage for security reason, so using index template to set policyID to manage index automatically will not be able anymore.

we add this ISM template to be the substitution for previous index template usage.

--- 

Bothered by a failing test which caused by Coordinator clear metadata which will be removed in next remove policyID PR, so comment out that test

Change Coordinator to implement ClusterStateListener but not LocalNodeMasterListener. If don't do this, I can see double logs. refer [alerting change](https://github.com/opendistro-for-elasticsearch/alerting/commit/7a09196b4bd4093ba7fd07e240e0f7a8bafbe399#diff-b3cb32ebb5bf1a6af7637b6374b5fd1eab7540e1ea4e17238dd8af4fcb76b70aR68)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/transport/action/indexpolicy/TransportIndexPolicyAction.kt [hunk 1]
```diff
@@ -79,6 +96,41 @@ class TransportIndexPolicyAction @Inject constructor(
...
+                    SearchSourceBuilder().query(
+                    QueryBuilders.existsQuery(ISM_TEMPLATE_FIELD)))
+                .indices(IndexManagementPlugin.INDEX_MANAGEMENT_INDEX)
+
+            client.search(searchRequest, object : ActionListener<SearchResponse> {
+                override fun onResponse(response: SearchResponse) {
+                    val ismTemplates = ismTemplatesFromSearchResponse(response, xContentRegistry)
+                    val overlaps = findConflictingISMTemplates(request.policyID, indexPatterns, priority, ismTemplates)
+                    if (overlaps.isNotEmpty()) {
+                        val esg = "new policy ${request.policyID} has an ism template with index pattern $indexPatterns " +
```
Comment (thalurur) [inline src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/transport/action/indexpolicy/TransportIndexPolicyAction.kt:117]: `esg` -> `msg`

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/transport/action/indexpolicy/TransportIndexPolicyAction.kt [hunk 2]
```diff
@@ -79,6 +96,41 @@ class TransportIndexPolicyAction @Inject constructor(
...
+                            "matching existing templates ${overlaps.entries.stream().map { "${it.key} => ${it.value}" }.collect(
```
Comment (thalurur) [inline src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/transport/action/indexpolicy/TransportIndexPolicyAction.kt:118]: templates -> policy?

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/transport/action/indexpolicy/TransportIndexPolicyAction.kt [hunk 3]
```diff
@@ -79,6 +96,41 @@ class TransportIndexPolicyAction @Inject constructor(
...
+            val searchRequest = SearchRequest()
+                .source(
+                    SearchSourceBuilder().query(
+                    QueryBuilders.existsQuery(ISM_TEMPLATE_FIELD)))
+                .indices(IndexManagementPlugin.INDEX_MANAGEMENT_INDEX)
+
+            client.search(searchRequest, object : ActionListener<SearchResponse> {
+                override fun onResponse(response: SearchResponse) {
+                    val ismTemplates = ismTemplatesFromSearchResponse(response, xContentRegistry)
+                    val overlaps = findConflictingISMTemplates(request.policyID, indexPatterns, priority, ismTemplates)
```
Comment (thalurur) [inline src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/transport/action/indexpolicy/TransportIndexPolicyAction.kt:115]: Just curious why is this is an issue - can we not assume if the priorities match for two policies containing matching indices the latest one is preferred since customer created it later?

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/ISMTemplateService.kt [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+         * @param indexMetadata cluster state index metadata
+         * @return policyID
+         */
+        @Suppress("ReturnCount")
+        fun findMatchingISMTemplate(ismTemplates: Map<String, ISMTemplate>, indexMetadata: IndexMetadata): String? {
+            val indexName = indexMetadata.index.name
+
+            // don't include hidden index
+            val isHidden = IndexMetadata.INDEX_HIDDEN_SETTING.get(indexMetadata.settings)
+            log.info("index $indexName is hidden $isHidden")
```
[the remark this answers is not in the document]
Comment (bowenlan-amzn) [reply src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/ISMTemplateService.kt:49]: removed this log, not important

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/ISMTemplateService.kt [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+            if (isHidden) return null
+
+            // only process indices created after template
+            // traverse all ism templates for matching ones
+            val patternMatchPredicate = { pattern: String -> Regex.simpleMatch(pattern, indexName) }
+            val matchedTemplates = mutableMapOf<ISMTemplate, String>()
+            ismTemplates.filter { (_, template) ->
+                log.info("index create after template? ${template.lastUpdatedTime.toEpochMilli() < indexMetadata.creationDate}")
```
[the remark this answers is not in the document]
Comment (bowenlan-amzn) [reply src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/ISMTemplateService.kt:57]: removed, ... should be cleaned up earlier

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/ISMTemplateService.kt [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+                template.lastUpdatedTime.toEpochMilli() < indexMetadata.creationDate
+            }.forEach { (templateName, template) ->
+                val matched = template.indexPatterns.stream().anyMatch(patternMatchPredicate)
+                if (matched) matchedTemplates[template] = templateName
```
[the remark this answers is not in the document]
Comment (bowenlan-amzn) [reply src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/ISMTemplateService.kt:61]: good point, changed

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/ManagedIndexCoordinator.kt [hunk 7]
```diff
@@ -261,10 +289,56 @@ class ManagedIndexCoordinator(
...
+        val indexToMatchedPolicy = indexNames.map { indexName ->
+            indexName to findMatchingISMTemplate(templates, indexMetadatas[indexName])
+        }.toMap()
+
+        val updateManagedIndexReqs = mutableListOf<DocWriteRequest<*>>()
+        indexToMatchedPolicy.filterNotNullValues()
+            .forEach { (index, policyID) ->
+            val indexUuid = indexMetadatas[index].indexUUID
+            if (indexUuid != null) {
+                logger.info("auto manage index $index to policy $policyID")
```
[the remark this answers is not in the document]
Comment (bowenlan-amzn) [reply src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/ManagedIndexCoordinator.kt:316]: done

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/ManagedIndexCoordinator.kt [hunk 8]
```diff
@@ -261,10 +289,56 @@ class ManagedIndexCoordinator(
...
+                    QueryBuilders.existsQuery(ISM_TEMPLATE_FIELD)))
+            .indices(INDEX_MANAGEMENT_INDEX)
+
+        return try {
+            val response: SearchResponse = client.suspendUntil { search(searchRequest, it) }
+            ismTemplatesFromSearchResponse(response).filterNotNullValues()
+        } catch (ex: IndexNotFoundException) {
+            emptyMap()
+        } catch (ex: ClusterBlockException) {
+            emptyMap()
```
[the remark this answers is not in the document]
Comment (bowenlan-amzn) [reply src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/ManagedIndexCoordinator.kt:338]: The reason I catch this here is because I see this exception when I run `./gradlew run` locally, the cluster is blocked for some time when it's being brought up.

For other situation, I am thinking if cluster is blocked, this whole ISM template function is also not functioning since there should not be any new indices created, so I just return a emptyMap()

will leave this conversation open for other reviewers to discuess

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/transport/action/indexpolicy/TransportIndexPolicyAction.kt (hunk 1 above)
Comment (bowenlan-amzn) [reply src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/transport/action/indexpolicy/TransportIndexPolicyAction.kt:117]: 😛 esg refer to error message

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/transport/action/indexpolicy/TransportIndexPolicyAction.kt (hunk 2 above)
Comment (bowenlan-amzn) [reply src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/transport/action/indexpolicy/TransportIndexPolicyAction.kt:118]: change to policy templates

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/transport/action/indexpolicy/TransportIndexPolicyAction.kt (hunk 3 above)
Comment (bowenlan-amzn) [reply src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/transport/action/indexpolicy/TransportIndexPolicyAction.kt:115]: I think this is just by design, we don't allow this. 
From the very beginning, we try to duplicate the usage of index template. Index template won't allow this.

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/ManagedIndexCoordinator.kt [hunk 9]
```diff
@@ -261,10 +289,56 @@ class ManagedIndexCoordinator(
             if (it.value.shouldDeleteManagedIndexMetaData()) indicesToRemoveManagedIndexMetaDataFrom.add(it.value.index)
         }
 
-        updateManagedIndices(updateManagedIndicesRequests + indicesDeletedRequests, hasCreateRequests)
+        // check if newly created indices matching any ISM templates
+        val updateMatchingIndexReqs = getMatchingIndicesUpdateReqs(event.state(), event.indicesCreated())
```
[the remark this answers is not in the document]
Comment (bowenlan-amzn) [reply src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/ManagedIndexCoordinator.kt:293]: changed

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/ISMTemplateService.kt [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+         * filter out hidden index
+         * filter out older index than template lastUpdateTime
+         *
+         * @param ismTemplates current ISM templates saved in metadata
+         * @param indexMetadata cluster state index metadata
+         * @return policyID
+         */
+        @Suppress("ReturnCount")
+        fun findMatchingISMTemplate(ismTemplates: Map<String, ISMTemplate>, indexMetadata: IndexMetadata): String? {
+            val indexName = indexMetadata.index.name
```
[the remark this answers is not in the document]
Comment (bowenlan-amzn) [reply src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/ISMTemplateService.kt:45]: good point, added

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/ManagedIndexCoordinator.kt [hunk 11]
```diff
@@ -261,10 +289,56 @@ class ManagedIndexCoordinator(
...
+        val indexMetadatas = clusterState.metadata.indices
+        val templates = getISMTemplates()
+
+        val indexToMatchedPolicy = indexNames.map { indexName ->
+            indexName to findMatchingISMTemplate(templates, indexMetadatas[indexName])
+        }.toMap()
+
+        val updateManagedIndexReqs = mutableListOf<DocWriteRequest<*>>()
+        indexToMatchedPolicy.filterNotNullValues()
+            .forEach { (index, policyID) ->
```
[the remark this answers is not in the document]
Comment (bowenlan-amzn) [reply src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/ManagedIndexCoordinator.kt:313]: yeah, I go through these function and renamed them.

**Review by thalurur — DISMISSED** [review]

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/elasticapi/ElasticExtensions.kt [hunk 12]
```diff
@@ -98,3 +106,26 @@ fun IndexMetadata.getManagedIndexMetaData(): ManagedIndexMetaData? {
...
+ */
+fun getPolicyToTemplateMap(response: SearchResponse, xContentRegistry: NamedXContentRegistry = NamedXContentRegistry.EMPTY):
+    Map<String, ISMTemplate?> {
+    return response.hits.hits.map {
+        val id = it.id
+        val seqNo = it.seqNo
+        val primaryTerm = it.primaryTerm
+        val xcp = XContentFactory.xContent(XContentType.JSON)
+            .createParser(xContentRegistry, LoggingDeprecationHandler.INSTANCE, it.sourceAsString)
+        xcp.parseWithType(id, seqNo, primaryTerm, Policy.Companion::parse)
```
Comment (dbbaughe) [inline src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/elasticapi/ElasticExtensions.kt:124]: This can throw an exception right?

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/ISMTemplateService.kt [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+            if (validationErrors.size > 0) {
+                val validationException = ValidationException()
+                validationException.addValidationErrors(validationErrors)
+                return IndexManagementException.wrap(validationException)
+            }
+            return null
+        }
+
+        /**
+         * find policy templates whose index patterns overlap with given template
```
Comment (dbbaughe) [inline src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/ISMTemplateService.kt:109]: What does overlap mean here? Just based on the index pattern?

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/util/IndexManagementException.kt [hunk 14]
```diff
@@ -0,0 +1,45 @@
+package com.amazon.opendistroforelasticsearch.indexmanagement.util
+
+import org.elasticsearch.ElasticsearchException
+import org.elasticsearch.common.Strings
+import org.elasticsearch.common.ValidationException
+import org.elasticsearch.index.IndexNotFoundException
+import org.elasticsearch.rest.RestStatus
+import java.lang.IllegalArgumentException
+
+class IndexManagementException(message: String, val status: RestStatus, ex: Exception) : ElasticsearchException(message, ex) {
```
Comment (dbbaughe) [inline src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/util/IndexManagementException.kt:10]: Are there any issues with serialization/deserialization of this exception class? I thought Bukhtawar or someone brought that up before.

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/transport/action/indexpolicy/TransportIndexPolicyAction.kt (hunk 1 above)
Comment (bowenlan-amzn) [reply src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/transport/action/indexpolicy/TransportIndexPolicyAction.kt:117]: changed

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/ISMTemplateService.kt [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+                    matchedPolicy = policyID
+                }
+            }
+
+            return matchedPolicy
+        }
+
+        /**
+         * validate the template Name and indexPattern provided in the template
+         * reusing ES validate function in MetadataIndexTemplateService
```
[the remark this answers is not in the document]
Comment (bowenlan-amzn) [reply src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/ISMTemplateService.kt:73]: Following logic [here](https://github.com/elastic/elasticsearch/blob/5852fbedf52c77b0a76c2c731fdd4982fb65cd87/server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java#L1244), not static function. And we only validate indexPatterns, we don't have template name anymore

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/model/ISMTemplate.kt [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+private val log = LogManager.getLogger(ISMTemplate::class.java)
+
+data class ISMTemplate(
+    val indexPatterns: List<String>,
+    val priority: Int,
+    val lastUpdatedTime: Instant
+) : ToXContentObject, Writeable {
+
+    init {
+        require(indexPatterns.isNotEmpty()) { "at least give one index pattern" }
```
[the remark this answers is not in the document]
Comment (bowenlan-amzn) [reply src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/model/ISMTemplate.kt:43]: change to 
```
require(priority >= 0) { "Requires priority to be >= 0" }
require(indexPatterns.isNotEmpty()) { "Requires at least one index pattern" }
```

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/model/ISMTemplate.kt [hunk 17]
```diff
@@ -0,0 +1,105 @@
...
+                xcp.nextToken()
+
+                when (fieldName) {
+                    INDEX_PATTERN -> {
+                        ensureExpectedToken(Token.START_ARRAY, xcp.currentToken(), xcp)
+                        while (xcp.nextToken() != Token.END_ARRAY) {
+                            indexPatterns.add(xcp.text())
+                        }
+                    }
+                    PRIORITY -> priority = if (xcp.currentToken() == Token.VALUE_NULL) 0 else xcp.intValue()
```
[the remark this answers is not in the document]
Comment (bowenlan-amzn) [reply src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/model/ISMTemplate.kt:92]: priority is not required, and we now require it to be >= 0

File: src/main/resources/mappings/opendistro-ism-config.json [hunk 18]
```diff
@@ -442,6 +442,23 @@
               }
             }
           }
+        },
+        "ism_template": {
+          "properties": {
+            "template_name": {
```
[the remark this answers is not in the document]
Comment (bowenlan-amzn) [reply src/main/resources/mappings/opendistro-ism-config.json:448]: good catch, this is stale now. Deleted

File: src/test/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/IndexStateManagementRestTestCase.kt [hunk 19]
```diff
@@ -709,4 +711,36 @@ abstract class IndexStateManagementRestTestCase : IndexManagementRestTestCase()
...
     }
+
+    protected fun assertPredicatesOnISMTemplatesMap(
+        templatePredicates: List<Pair<String, List<Pair<String, (Any?) -> Boolean>>>>, // response map name: predicate
+        response: Map<String, Any?>
+    ) {
+        val templates = response["ism_templates"] as ArrayList<Map<String, Any?>>
+
+        templatePredicates.forEachIndexed { ind, (_, predicates) ->
+            // assertTrue("The template: $name was not found in the response: $response", templates.containsKey(name))
```
[the remark this answers is not in the document]
Comment (bowenlan-amzn) [reply src/test/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/IndexStateManagementRestTestCase.kt:722]: not actually. seems this is just because template name is not there anymore.

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/ISMTemplateService.kt (hunk 13 above)
Comment (bowenlan-amzn) [reply src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/ISMTemplateService.kt:109]: changed to `conflictingPolicyTemplates`
yeah, it checks if the new template index pattern matching range overlaps with any old template index pattern

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/ISMTemplateService.kt [hunk 20]
```diff
@@ -0,0 +1,137 @@
...
+import org.apache.lucene.util.automaton.Operations
+import org.elasticsearch.ElasticsearchException
+import org.elasticsearch.cluster.metadata.IndexMetadata
+import org.elasticsearch.common.Strings
+import org.elasticsearch.common.ValidationException
+import org.elasticsearch.common.regex.Regex
+
+private val log = LogManager.getLogger(ISMTemplateService::class.java)
+
+class ISMTemplateService {
```
[the remark this answers is not in the document]
Comment (bowenlan-amzn) [reply src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/ISMTemplateService.kt:31]: change to extension function, feeling good

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/ManagedIndexCoordinator.kt [hunk 21]
```diff
@@ -261,10 +289,56 @@ class ManagedIndexCoordinator(
             if (it.value.shouldDeleteManagedIndexMetaData()) indicesToRemoveManagedIndexMetaDataFrom.add(it.value.index)
         }
 
-        updateManagedIndices(updateManagedIndicesRequests + indicesDeletedRequests, hasCreateRequests)
+        // check if newly created indices matching any ISM templates
+        val updateMatchingIndexReq = getMatchingIndicesUpdateReq(event.state(), event.indicesCreated())
+        if (updateMatchingIndexReq.isNotEmpty()) hasCreateRequests = true
+
+        updateManagedIndices(updateManagedIndicesRequests + updateMatchingIndexReq + indicesDeletedRequests, hasCreateRequests)
```
[the remark this answers is not in the document]
Comment (bowenlan-amzn) [reply src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/ManagedIndexCoordinator.kt:296]: we scan all indices policyID setting before finding matching index, I can record all indices with policyID, remove them from indicesCreated() if exist.

[the remark this answers is not in the document]
Comment (dbbaughe) [reply src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/ManagedIndexCoordinator.kt:296]: Yes, as of now you'll end up with two create requests for the same index right? SInce we're not removing the setting in this PR let's do like you said and make sure the templates skip any indices that have the setting on them (prefer explicitness). I would imagine it's shortly removed after with the follow up PR though which is fine.

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/ISMTemplateService.kt (hunk 13 above)
Comment (dbbaughe) [reply src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/ISMTemplateService.kt:109]: Hmm is the behavior correct? I see it used below to reject a PutPolicy request, so if there is overlap in index patterns it rejects it? I thought we allow overlap in index patterns, which is why we have the priority to control which one takes priority?

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/ManagedIndexCoordinator.kt [hunk 22]
```diff
@@ -261,10 +289,56 @@ class ManagedIndexCoordinator(
...
         clearManagedIndexMetaData(indicesToRemoveManagedIndexMetaDataFrom)
     }
 
+    /**
+     * build requests to create jobs for indices matching ISM templates
+     */
+    suspend fun getMatchingIndicesUpdateReq(clusterState: ClusterState, indexNames: List<String>): List<DocWriteRequest<*>> {
+        val indexMetadatas = clusterState.metadata.indices
+        val templates = getISMTemplates()
```
[the remark this answers is not in the document]
Comment (bowenlan-amzn) [reply src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/ManagedIndexCoordinator.kt:305]: changed

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/ManagedIndexCoordinator.kt [hunk 23]
```diff
@@ -261,10 +289,56 @@ class ManagedIndexCoordinator(
...
+    }
+
+    suspend fun getISMTemplates(): Map<String, ISMTemplate> {
+        val searchRequest = SearchRequest()
+            .source(
+                SearchSourceBuilder().query(
+                    QueryBuilders.existsQuery(ISM_TEMPLATE_FIELD)))
+            .indices(INDEX_MANAGEMENT_INDEX)
+
+        return try {
```
[the remark this answers is not in the document]
Comment (bowenlan-amzn) [reply src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/ManagedIndexCoordinator.kt:332]: yes, added a general exception catch
```
catch (e: Exception) {
    logger.error("Failed to get ISM templates", e)
    emptyMap()
}
```

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/ISMTemplateService.kt (hunk 13 above)
Comment (bowenlan-amzn) [reply src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/ISMTemplateService.kt:109]: overlap index pattern is allowed when priority is different, if priority is same, putPolicy is blocked

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/model/ISMTemplate.kt [hunk 24]
```diff
@@ -0,0 +1,105 @@
...
+
+    init {
+        require(indexPatterns.isNotEmpty()) { "at least give one index pattern" }
+    }
+
+    override fun toXContent(builder: XContentBuilder, params: ToXContent.Params): XContentBuilder {
+        return builder.startObject()
+            .field(INDEX_PATTERN, indexPatterns)
+            .field(PRIORITY, priority)
+            .optionalTimeField(LAST_UPDATED_TIME_FIELD, lastUpdatedTime)
```
[the remark this answers is not in the document]
Comment (bowenlan-amzn) [reply src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/model/ISMTemplate.kt:50]: I am taking Policy as [example](https://github.com/opendistro-for-elasticsearch/index-management/blob/8c7067ce95ddb5c842cdaec528c575478d34dfe9/src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/model/Policy.kt#L72)

Theoretically, everytime we parse the JSON, this field is updated.

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/util/IndexManagementException.kt (hunk 14 above)
Comment (bowenlan-amzn) [reply src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/util/IndexManagementException.kt:10]: for the IndexManagementException, currently is only used like this `actionListener.onFailure(...)`
so I think it doesn't involve inter-node communication

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/elasticapi/ElasticExtensions.kt (hunk 12 above)
Comment (bowenlan-amzn) [reply src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/elasticapi/ElasticExtensions.kt:124]: used in:
1. Coordinator: catch in the general exception, logger out the error, return empty
2. TransportIndexPolicyAction: directly throw `IllegalArgumentException`

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/ManagedIndexCoordinator.kt (hunk 8 above)
[the remark this answers is not in the document]
Comment (bowenlan-amzn) [reply src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/ManagedIndexCoordinator.kt:346]: wrapped getISMTemplates call within newly created indices event

File: src/main/resources/mappings/opendistro-ism-config.json [hunk 25]
```diff
@@ -442,6 +442,20 @@
               }
             }
           }
+        },
+        "ism_template": {
```
Comment (qreshi) [inline src/main/resources/mappings/opendistro-ism-config.json:446]: Do we need to increment the schema of the mappings or has it already been incremented after the last release?

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/ISMTemplateService.kt [hunk 26]
```diff
@@ -0,0 +1,131 @@
...
+): Map<String, List<String>> {
+    val automaton1 = Regex.simpleMatchToAutomaton(*indexPatterns.toTypedArray())
+    val overlappingTemplates = mutableMapOf<String, List<String>>()
+
+    // focus on template with same priority
+    this.filter { it.value.priority == priority }
+        .forEach { (policyID, template) ->
+        val automaton2 = Regex.simpleMatchToAutomaton(*template.indexPatterns.toTypedArray())
+        if (!Operations.isEmpty(Operations.intersection(automaton1, automaton2))) {
+            log.info("existing ism_template in $policyID overlaps candidate $candidate")
```
Comment (qreshi) [inline src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/ISMTemplateService.kt:124]: NP: Can reword this to "Existing ism_template `for` $policyID..."

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/ISMTemplateService.kt [hunk 27]
```diff
@@ -0,0 +1,131 @@
...
+    priority: Int
+): Map<String, List<String>> {
+    val automaton1 = Regex.simpleMatchToAutomaton(*indexPatterns.toTypedArray())
+    val overlappingTemplates = mutableMapOf<String, List<String>>()
+
+    // focus on template with same priority
+    this.filter { it.value.priority == priority }
+        .forEach { (policyID, template) ->
+        val automaton2 = Regex.simpleMatchToAutomaton(*template.indexPatterns.toTypedArray())
+        if (!Operations.isEmpty(Operations.intersection(automaton1, automaton2))) {
```
Comment (qreshi) [inline src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/ISMTemplateService.kt:123]: Just to make sure, since this is an intersection of automatons, it will cover cases where one pattern is a subset of another right?

For example, say one template is created with pattern `exa*` and another is created with `exam*` both with the same priority. That case should considered a conflict.

Could we also add a test for this case. I see there is one for overlapping patterns when the pattern is the exact same but this particular case would be good to capture as well.

File: src/main/resources/mappings/opendistro-ism-config.json (hunk 25 above)
Comment (bowenlan-amzn) [reply src/main/resources/mappings/opendistro-ism-config.json:446]: good point, I think we can increase the version to 7 in the security PR.

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/ISMTemplateService.kt (hunk 26 above)
Comment (bowenlan-amzn) [reply src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/ISMTemplateService.kt:124]: reworded

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/ISMTemplateService.kt (hunk 15 above)
[the remark this answers is not in the document]
Comment (bowenlan-amzn) [reply src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/ISMTemplateService.kt]: rewrite a little bit and acknowledge the original contributor

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/ISMTemplateService.kt (hunk 27 above)
Comment (bowenlan-amzn) [reply src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/ISMTemplateService.kt:123]: changed that test to have 2 templates with index patterns `log*` and `lo*`

**Review by dbbaughe — APPROVED** [review]

**Review by qreshi — APPROVED** [review]
